src/lib/api/prefetch.ts

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (c) 2018 by Filestack.
+ * Some rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ClientOptions, Session } from '../client';
+import { PickerOptions } from './../picker';
+import { requestWithSource } from '../api/request';
+
+type PrefetchOptionsSetting = {
+  inapp_browser: boolean;
+  customsource: boolean;
+};
+
+type PrefetchOptionsPermissions = {
+  intelligent_ingestion: boolean;
+  blocked: boolean;
+  blacklisted: boolean;
+  whitelabel: boolean;
+};
+
+type PrefetchOptions = {
+  pickerOptions: PickerOptions;
+  settings: PrefetchOptionsSetting;
+  permissions: PrefetchOptionsPermissions;
+};
+
+type responseObject = {
+  intelligent_ingestion: boolean;
+  blocked: boolean;
+  blacklisted: boolean;
+  whitelabel: boolean;
+  customsource: boolean;
+  inapp_browser: boolean;
+  updated_config: {
+    fromSources: string[];
+  };
+};
+
+/**
+ * @private
+ */
+export class Prefetch {
+  private session: Session;
+
+  private prefetchUrl: string;
+
+  private configToCheck: PickerOptions;
+
+  constructor(session: Session) {
+    this.session = session;
+    this.prefetchUrl = session.urls.cloudApiUrl;
+  }
+
+  async getConfig({ pickerOptions, settings, permissions }: PrefetchOptions) {
+    // @todo: stay?
+    if (this.session.prefetch) {
+      return Promise.resolve(this.session.prefetch);
+    }
+
+    const configToSend = this.cleanUpCallback(pickerOptions);
+
+    const paramsToSend = {
+      apikey: this.session.apikey,
+      security: {
+        policy: this.session.policy,
+        signature: this.session.signature,
+      },
+      settings,
+      permissions,
+      picker_config: configToSend,
+    };
+
+    const response = await requestWithSource()
+      .post(`${this.prefetchUrl}/prefetch`, paramsToSend)
+      .then(res => res.data)
+      .then(data => data);
+
+    return this.reassignCallbacks(response);
+  }
+
+  private cleanUpCallback(pickerOptions: PickerOptions) {
+    const tempCallbackVariable: null = null;
+
+    this.configToCheck = pickerOptions;
+
```

Comment:
In this case You have to do (easiest way)
1. make copy of config (u cant have referance) You can sue https://lodash.com/docs/4.17.15#cloneDeep
2. cleanup all functions from config copy
3. return copied config for request
4. after the received response from backend, merge old config with new one (in that case callbacks will stay the same) - ie u can do it by using one of https://mariusschulz.com/blog/combining-settings-objects-with-lodash-assign-or-merge